### PR TITLE
dbNSFP v4.0a

### PR DESCRIPTION
**Pull request by matthdsm**
Hi Brad,

I've updated the GGD recipe's to fetch the latest version of dbNSFP and updated the logic for GRCh37 a bit.

I also cleaned up some leftover debug scripts for gnomAD.

Cheers
M

### CODE REVIEW

**Comment by chapmanb** [comment]
Matthis;
Thanks much for keeping this up to date. I'm always a bit nervous on updating dbNSFP because of the huge file sizes but if this is working for you and useful happy to have you supporting it. It looks like this PR conflicts with Sergey's recent changes and is changing/deleting some of the gnomad scripts. Is it possible to get a clean PR with just the dbNSFP updates? Thanks again.

**Comment by matthdsm** [comment]
Hi Brad,

Should be good to go now! Fair warning, the processing time and system resources for this are quite big. During the sort, the files expanded to about 120GB and it took the better part of a day to get it finished.
The biggest issue I have with this version is that the zipped file contains gzipped files per chromosome, instead of plain text files as was the case in previous versions. The addition of gnomAD exomes/genomes grew the file size quite a bit.

I do think the preprocessing could still be streamlined and made more efficient, but I haven't got the time to dig into it.

Cheers
M

**Comment by chapmanb** [comment]
Matthias;
Thanks much for the PR cleanup and the details about the run. Are the 4.0 updates worth the extra processing and preparation time? I worry about this being too slow for general use and updates. If there is a compelling reason to move to 4.0 happy to do it if users will understand the extra improvements they'll get for the switch.

**Comment by matthdsm** [comment]
Not much is "new", but some predictions have been added, and other have been updated, so I do think it's worthwhile. However, I'm a bit of a sucker for staying up to date on data resources, so it might be just me..

M

**Comment by matthdsm** [comment]
The biggest addition here is the gnomAD update, but that's kinda useless for us, since we source that data straight from gnomAD itself. 
Also, I trust the liftover from Ensembl more, than the one in dbNSPF, since I couldn't really find much documentation on how they did it, whereas Ensembl has documented their entire flow.

**Comment by chapmanb** [comment]
Matthias;
Thanks for the details. It sounds from your description like there aren't many practical advantages to updating in bcbio and the preparation step is much slower. Do you think we should still do this? It seems like we're adding work since we'll need to support and improve this, without a perceived benefit for the users. I'd be inclined not to change things that are working now without a good reason. What do you think?

**Comment by matthdsm** [comment]
I'd say go for it. However, that may be my personal bias.
As I see it, there won't be much difference in support cost, except for an issue here and there. If needed, you can ping me for that.
I also expect other users to ask for an update, as there already have been issues requesting this, even when 4.0 was still in beta.
New users will have to deal with the processing time either way, upgrade or not.
Existing users will take a hit, but with the advantage of the updated annotations. People not interested in what dbNSFP has to offer can leave out the `datatarget` and won't ever notice.

Of course, as I said, I'm probably biased, so I'm leaving it up to you.

M

**Comment by naumenko-sa** [comment]
Hi Brad and Matthias!

From dbNFSP I am using 6 fields:
```
#dbNSFP v3.4
[[annotation]]
file = "variation/dbNSFP.txt.gz"
names = ["CADD_phred","phyloP20way_mammalian","phastCons20way_mammalian","Vest3_score","Revel_score","Gerp_score"]
columns = [79,111,115,58,70,107]
ops = ["first","first","first","first","first","first"]
```
So I'm personally and colleagues doing similar analysis won't benefit from the upgrade, which is mostly duplidated VEP and Gnomad info in dbSNP.  dbNSFP was one of the slowest part of bcbio to install and many won't be happy to touch it. 

Now I'm even not sure if we should update to gnomad2.1 from 2.0.5 where the benefits are more obvious.

I'd propose a compromise solution, similar to gnomad2.1 approach: nobody needs all 300+ fields from dbNSFP. I, for example, need 6. There might be a reasonable default number of fields to install (no gnomad, no vep etc), just ~10 most popular fields like polyphen, gerp and cadd. There should be an instruction on how to update if anybody wants more fields (just modifying the recipe).
In that way, we will be formally in a sync with dbNFSP, and installation will be much easier and resulting file much smaller to the benefit of everyone.

Sergey

**Comment by matthdsm** [comment]
I can roll with the dbNSFP "light" version, however, I think only supporting a light version with essentials will:
a) dissapoint a lot of (new) people expecting the full dataset
b) open another whole can of worms about which fields to include in the "light" version. For example, in the vcfanno config I've already excluded all fields deemed unimportant and all fields for which we've already got data from other sources (VEP/snpEff/gnomad/....)

In your case, that would mean you'd lose the CADD score, since that's already included in the VEP annotations. As it is now, users can select required data by just editing the config file. When trimming the file, that would no longer be possible, and users would have to download the full version anyway.

so TL;DR
I'm okay with a limited version, but there needs to be a genuine discussion on which fields to remove, since the usecases for dbNSFP can vary greatly.

Cheers
M

**Comment by matthdsm** [comment]
Also, I suppose there is a way to streamline the dbNSFP preprocessing.
There is no reason at all why the files per chromosome are required to be merged. As it is, I think we could just use the separate files and edit the vcfanno config. 
In that case, all we'd have to do is unpack the zip file (contains gzipped tabular files) and create indexes for each file.

How's that for you?

Cheers
M

**Comment by matthdsm** [comment]
I just did some quick test wrt processing time.
Unpacking and re-bgzipping one chr takes about 10 min, and 3 min for indexing.
I guess that processing the entire set would take about 5h, excluding download times.
This is a huge improvement already, compared to the 12h+ it took for the merged version.
Too bad the guys at jpopgen didn't use bgzip to package the data, otherwise the entire set would be processed in about 30min.

M

**Comment by naumenko-sa** [comment]
Hi Matthias!

Thanks for being on top of dbNSFP update issue! Sorry that it took so long to review this update.

Unfortunately, after several rounds of consideration, it is hard to convince hundreds of imaginary owners of existing bcbio installations that this update would be worth the efforts:

Cons:
- installing dbNSFP as multiple file collection is not a good option, as it would break existing scripts, and unnecessarily complicate the annotation;
- we don't really need Gnomad, snpEff, and VEP annotations from dbNSFP, because we have the originals;
- installing dbNSFP4.0 as a single file will require >12hours and >100GB of space in the process; 
For many users supporting two references (hg19/hg39) that would mean >1day of bcbio downtime;
- dbNSFP4.0 column order is different from dbNSFP3.5 (the same score would be in a different column). That might lead to bugs which would be hard to trace. Users would need to update their annotation scripts, otherwise, they would get incorrect annotations.

Pros:
- some new scores (Deogen, PrimateAI)?

So we'd keep dbNSFP version 3.5a for now.

Thanks again for your multiple contributions to cloudbiolinux/bcbio and being the driver of many improvements!

Sergey

**Comment by matthdsm** [comment]
Hi Sergey,

No biggie!


321i3d

Cheers
M

**Comment by matthdsm** [comment]
closes bcbio/bcbio-nextgen#2820
